packages/flutter/lib/src/painting/notched_shapes.dart

```diff
@@ -0,0 +1,106 @@
+import 'dart:math' as math;
+
+import 'package:flutter/rendering.dart';
+
+/// A shape with a notch in its outline.
+///
+/// Typically used as the outline of a 'host' widget to make a notch that
+/// accomodates a 'guest' widget. e.g the [BottomAppBar] may have a notch to
+/// accomodate the [FloatingActionBar].
+abstract class NotchedShape {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const NotchedShape();
+
+  /// Creates a [Path] that describes the outline of the shape.
+  ///
+  /// The `host` is the bounding rectangle of the shape.
+  ///
+  /// Rhe `guest` is the bounding rectangle of the shape for which a notch will
+  /// be made.
+  Path getOuterPath(Rect host, Rect guest);
+}
+
+/// A rectangle with a smooth circular notch.
+class CircularNotchedRectangle implements NotchedShape {
```

Comment:
consider calling this something like "DimpleNotchedRectangle" or some such, to make it clearer what the difference is between this and a semicircular notched rectangle.

Reply:
Naming here is pretty hard...
For a few moments I thought that once we expose the curve computation parameters (s1 and s2) we will avoid this naming problem, as one could configure this to be equivalent to the single arc notch, but there's actually no such setting for s1 and s2.

Another option would be to add a `smooth` boolean parameter to this class, but that's also not ideal as when it's false the s1/s2 parameters (which I plan to expose anyway) will be meaningless.

@HansMuller maybe you have some thoughts here?

Reply:
Maybe generate a simple circular notch if the s1/s2 parameters are null but have them default to values that produce a dimple.

Reply:
I'm ok with that - if either s1 or s2 are null we'll just default to the single arc notch. An this thing stays the CircularNotchedRectangle.

Reply:
Shouldn't there be s1/s2 constructor parameters?